### PR TITLE
[OpenMP] Add support for lowering Critical Construct

### PR DESCRIPTION
**Pull request by kiranchandramohan**
The patch adds lowering for the Critical Construct.

The patch uses some code from @Sameeranjoshi's patch https://reviews.llvm.org/D93051.

### CODE REVIEW

**Review by jeanPerier — COMMENTED** [review]
> Is that OK? Also what should be the mangling if any?

My concern would be potential conflicts with other program symbols after remangling (for instance, a `BIND(C, name="help")` variable somewhere else). Even if it is illegal as you are pointing out, I am not sure the front-end enforces it, so adding a bit more of protection seems safer to me and would also help navigating through the globals/symbols in  the object files.

Is clang also creating a global of a certain name for the C/C++ equivalent program if there is any ?

File: flang/lib/Lower/OpenMP.cpp
Comment (schweitzpgi) [inline flang/lib/Lower/OpenMP.cpp:371]: It looks like `else` should be deleted.

File: flang/lib/Lower/OpenMP.cpp
Comment (schweitzpgi) [reply flang/lib/Lower/OpenMP.cpp:608]: You could even use a lambda to initialize it directly (if needed), as in

```c++
auto criticalOp = [&]() -> mlir::omp::CriticalOp { ... }();
```

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (schweitzpgi) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:238]: Yes. The assumption is that upstreamed changes are more correct than whatever was cherry-picked in conflicts other than lowering and optimization.

**Review by schweitzpgi — COMMENTED** [review]
On the mangling question: this can use a simple bare name, the unidirectional mangling (`_QQ...`), just "pretend" these are global constants, or we could add a different bijective mangling for this case. Personally, I'd lean towards the last in order to avoid any unpleasant compiler crashes due to a possible IR symbol binding collision.

File: flang/lib/Lower/OpenMP.cpp
Comment (kiranchandramohan) [reply flang/lib/Lower/OpenMP.cpp]: Thanks, I have used @schweitzpgi's suggestion.

File: flang/lib/Lower/OpenMP.cpp
Comment (kiranchandramohan) [reply flang/lib/Lower/OpenMP.cpp:629]: Good point. Yes, we can break. There will only be one hint clause and the standard prescribes this through the syntax.
```
!$omp critical [(name) [[,] hint(hint-expression)] ]
```
I have added a change to the OMP.td file to enforce this restriction.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (kiranchandramohan) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:238]: :) We need John the merciful.

I think we have now added a few changes to the mlir repo, I will upstream all changes possible.

File: flang/lib/Semantics/resolve-directives.cpp
Comment (kiranchandramohan) [reply flang/lib/Semantics/resolve-directives.cpp:1470]: That patch has been lying idle for a while since @Sameeranjoshi moved to another team. But I hear that he is back. So there is a good chance that this will land soon with some changes (including this change). Otherwise, I will take the patch forward.

File: flang/lib/Lower/OpenMP.cpp
Comment (kiranchandramohan) [reply flang/lib/Lower/OpenMP.cpp]: Assuming this is a coding style comment. Done.
https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

**Comment by kiranchandramohan** [comment]
> My concern would be potential conflicts with other program symbols after remangling (for instance, a BIND(C, name="help") variable somewhere else). Even if it is illegal as you are pointing out, I am not sure the front-end enforces it, so adding a bit more of protection seems safer to me and would also help navigating through the globals/symbols in the object files.

You are suggesting a mangling for openmp critical here, right?

> Is clang also creating a global of a certain name for the C/C++ equivalent program if there is any ?

Yes. For e.g. Clang creates the following for a critical section with name "nm1".
@.gomp_critical_user_nm1.var = common global [8 x i32] zeroinitializer

> On the mangling question: this can use a simple bare name, the unidirectional mangling (_QQ...), just "pretend" these are global constants, or we could add a different bijective mangling for this case. Personally, I'd lean towards the last in order to avoid any unpleasant compiler crashes due to a possible IR symbol binding collision.

OK. I will try to add a different mangling.

I have another question/point for discussion here. We currently create the following global in FIR for a critical section with name (for eg help).
```
fir.global @help constant : !fir.char<1,4>
```
This gets converted into the following in LLVM IR.
```
@help = external constant [4 x i8]
```
The OpenMP IRBuilder also creates the following for use in the runtime call below. Now `@help = external constant [4 x i8]` in LLVM IR created from the fir.global is probably extraneous and irrelevant. But at the same time, the fir.global is necessary to enforce the naming issue in FIR. 
```
@.gomp_critical_user_.var = common global [8 x i32] zeroinitializer
  call void @__kmpc_critical_with_hint(%struct.ident_t* @1, i32 %omp_global_thread_num, [8 x i32]* @.gomp_critical_user_help.var, i32 2), !dbg !10
  ```
  
  Is it possible to just declare a name in FIR to represent a global (like a function declaration) but not create a variable (because this will ultimately get created in LLVM IR)?

**Comment by kiranchandramohan** [comment]
ping.

**Comment by jeanPerier** [comment]
> You are suggesting a mangling for openmp critical here, right?

Yes, I think it is safer.

> Is it possible to just declare a name in FIR to represent a global (like a function declaration) but not create a variable (because this will ultimately get created in LLVM IR)?

Do you mean to somehow reserve the name without creating any fir.global ? If so, I do not think it is possible because the IR itself is used for the bookeeping of fir.global. Couldn't the OpenMP IRBuilder somehow recognized the global created for the critical and replace it or delete it (or maybe the mlir::omp::CriticalOp rewrite in MLIR )?

**Comment by schweitzpgi** [comment]
> Is it possible to just declare a name in FIR to represent a global (like a function declaration) but not create a variable (because this will ultimately get created in LLVM IR)?

I'm not entirely sure what you're looking for. It is possible to create globals with different linkages and/or without initializers. There are some restrictions though from what the LLVM-IR dialect of MLIR will support or not, so maybe that is what you're experiencing?

**Comment by kiranchandramohan** [comment]
> > Is it possible to just declare a name in FIR to represent a global (like a function declaration) but not create a variable (because this will ultimately get created in LLVM IR)?
> 
> I'm not entirely sure what you're looking for. It is possible to create globals with different linkages and/or without initializers. There are some restrictions though from what the LLVM-IR dialect of MLIR will support or not, so maybe that is what you're experiencing?

Thanks @schweitzpgi for your question. Ultimately I decided to handle this in the OpenMP dialect itself. In the MLIR layer, we need names of OpenMP Critical constructs to be unique. This is handled by creating a "declare critical" operation that will declare names of critical sections. The existence of such an operation also removes the need for mangling. The OpenMP IRBuilder will create a variable (with a name derived from the critical section name) that functions as a lock for the critical section. We only need to pass the name of the critical section to the OpenMP IRBuilder.
This is approved upstream (https://reviews.llvm.org/D108713). I will cherry-pick it here after landing.

**Comment by schweitzpgi** [comment]
> Thanks @schweitzpgi for your question. Ultimately I decided to handle this in the OpenMP dialect itself. In the MLIR layer, we need names of OpenMP Critical constructs to be unique. This is handled by creating a "declare critical" operation that will declare names of critical sections. The existence of such an operation also removes the need for mangling. The OpenMP IRBuilder will create a variable (with a name derived from the critical section name) that functions as a lock for the critical section. We only need to pass the name of the critical section to the OpenMP IRBuilder.
> This is approved upstream (https://reviews.llvm.org/D108713). I will cherry-pick it here after landing.

Thanks for the explanation, @kiranchandramohan. Makes sense. I assume the OpenMP IRBuilder naming scheme won't conflict with ad hoc Fortran naming schemes that might be implemented for interoperating with other Fortran compilers, etc.

File: flang/lib/Lower/OpenMP.cpp
Comment (kiranchandramohan) [reply flang/lib/Lower/OpenMP.cpp]: I have made this change as the Critical function lowering has become a bit bigger.

Please note that this is only suggestive and we usually stick to mlir/llvm coding guidelines in the Lower and Optimize directories.

File: flang/test/Lower/OpenMP/critical.f90
Comment (kiranchandramohan) [reply flang/test/Lower/OpenMP/critical.f90:26]: Done.

File: mlir/test/Target/LLVMIR/openmp-llvm.mlir
Comment (kiranchandramohan) [reply mlir/test/Target/LLVMIR/openmp-llvm.mlir:509]: Done.

**Comment by kiranchandramohan** [comment]
> > Thanks @schweitzpgi for your question. Ultimately I decided to handle this in the OpenMP dialect itself. In the MLIR layer, we need names of OpenMP Critical constructs to be unique. This is handled by creating a "declare critical" operation that will declare names of critical sections. The existence of such an operation also removes the need for mangling. The OpenMP IRBuilder will create a variable (with a name derived from the critical section name) that functions as a lock for the critical section. We only need to pass the name of the critical section to the OpenMP IRBuilder.
> > This is approved upstream (https://reviews.llvm.org/D108713). I will cherry-pick it here after landing.
> 
> Thanks for the explanation, @kiranchandramohan. Makes sense. I assume the OpenMP IRBuilder naming scheme won't conflict with ad hoc Fortran naming schemes that might be implemented for interoperating with other Fortran compilers, etc.

Unlikely to conflict with Fortran names since there will be a `gomp` and `critical` prefix, e.g. for a critical section with name `help` the variable will have the name `@.gomp_critical_user_help.var`.

I have now cherry-picked https://reviews.llvm.org/D108713 (and quite a few dependent commits to make the build clean).

Have also addressed all review comments.

File: flang/lib/Lower/OpenMP.cpp
Comment (PeixinQiao) [reply flang/lib/Lower/OpenMP.cpp]: Thanks for the explanation.

**Comment by kiranchandramohan** [comment]
ping. Can this be approved if OK?

**Review by clementval — APPROVED** [review]

**Comment by kiranchandramohan** [comment]
The first 15 commits in the PR are cherry-picked from upstream llvm-project. Would it be better to have these cherry-picked commits in a separate PR for ease of review and approval?

Also, in general can we have a simplified process for approving and merging cherry-picked commits from upstream llvm-project?

**Comment by jeanPerier** [comment]
#1055 has been merged, can this one be rebased ?

**Comment by kiranchandramohan** [comment]
Thanks for the ping @jeanPerier. Rebased.

**Review by jeanPerier — APPROVED** [review]
Thanks for rebasing @kiranchandramohan .
Lowering logs good to me. Please make sure to upstream the parser/semantics changes, it looks like they are not yet in LLVM.

**Comment by kiranchandramohan** [comment]
Sure @jeanPerier. AMD folks are working on the parser/semantics part in upstream llvm. I will ask them to cherry-pick it here and solve any conflicts.

@schweitzpgi can this be merged?

**Pull request merged**
